### PR TITLE
ec2_group: add ports keyword (#19888).

### PR DESCRIPTION
**Pull request by anryko**
##### ISSUE TYPE

 - Feature Pull Request

##### COMPONENT NAME

ec2_group

##### ANSIBLE VERSION

```
ansible 2.3.0 (ec2_group_ports 460d8f34a9) last updated 2017/01/29 17:29:54 (GMT +200)
```

##### SUMMARY

Added 'ports' rule keyword. It accepts a single port value or a list of values including ranges (from_port-to_port). This will reduce verbosity of rules definition, increase maintenance convenience and readability of complex ec2 security groups. 

```
- name: example2 ec2 group
  ec2_group:
    name: example2
    description: an example2 EC2 group
    vpc_id: 12345
    region: eu-west-1a
    rules:
      - proto: tcp
        ports: 22
        group_name: example-vpn
      - proto: tcp
        ports:
          - 80
          - 443
          - 8080-8099
        cidr_ip: 0.0.0.0/0
```

### CODE REVIEW

**Comment by jbscare** [comment]
I haven't looked at the code, but I like the YAML side.

File: lib/ansible/modules/cloud/amazon/ec2_group.py
Comment (willthames) [inline lib/ansible/modules/cloud/amazon/ec2_group.py]: If `ports_expand` took a rule rather than the value of its `ports` key this entire function could be replaced with

```
return [ports_expand(rule) for rule in rules]
```

Edit: of course `ports_expand` returns a list, so you need to flatten the result:

```
return [rule for expansion in ports_expand(rule) for rule in expansion]
```

File: lib/ansible/modules/cloud/amazon/ec2_group.py
Comment (willthames) [inline lib/ansible/modules/cloud/amazon/ec2_group.py]: typo: string

**Comment by willthames** [comment]
Thanks for the fixes - I will test this today.

**Comment by willthames** [comment]
shipit

File: lib/ansible/modules/cloud/amazon/ec2_group.py
Comment (willthames) [inline lib/ansible/modules/cloud/amazon/ec2_group.py]: This is great, I really like this.

Can you make it clearer that all source types are now supported (group_name, group_id and cidr_ip) - it's clear to me from reading the source code, but wasn't when I looked at this example (might be worth having separate examples for different source types?)

File: lib/ansible/modules/cloud/amazon/ec2_group.py
Comment (willthames) [inline lib/ansible/modules/cloud/amazon/ec2_group.py]: imports should all be just after the documentation sections (effectively at the top of the actual non-doc code)

**Review by willthames — CHANGES_REQUESTED** [review]

**Comment by willthames** [comment]
I've removed my earlier shipit comment as more functional changes have been made - I'll retest with the new features.

**Comment by anryko** [comment]
@willthames is there anything else I can improve?

**Comment by willthames** [comment]
@anryko I don't think so - I just need to find some time to test the most recent enhancements - they look good at first glance.

**Comment by alexlo03** [comment]
Consider adding an integration test at https://github.com/ansible/ansible/blob/devel/test/integration/roles/test_ec2_group/tasks/main.yml ?

**Comment by willthames** [comment]
shipit

Have tested this in our environment (which has SGs that very much benefit from this change) and it appears to work perfectly.

@alexlo03's suggestion about adding tests is a good one, but I'm not going to block this on that.

**Comment by anryko** [comment]
@alexlo03 as I understand tests are done against AWS API responses and nothing was changed in this regard. Ansible task syntax was enriched, but underneath it is expanded to meet the same API as before.
Thanks for pointing this out, I will make sure to check integration tests while working on future commits.

**Comment by willthames** [comment]
!needs_revision

**Comment by willthames** [comment]
bot_status

**Comment by jctanner** [comment]
@willthames you need to dismiss or approve your review to get this out of needs revision.
```
changes_requested_by: willthames
```

**Review by willthames — APPROVED** [review]

**Review by s-hertel — APPROVED** [review]

**Comment by s-hertel** [comment]
shipit

**Pull request merged**
